### PR TITLE
First Draft of SQL Guidelines

### PR DESCRIPTION
**Pull request by rr-manoj-dobbali**
This guide is created to standardize the sql scripting style across projects/teams.

### CODE REVIEW

**Comment by rr-yun-liaw** [comment]
test

File: README.md [hunk 1]
```diff
@@ -1,2 +1,462 @@
...
+-- Bad
+select count(*) as itemcount
+```
+
+- For Indentation, use 4 spaces for one level of indentation and 8 for deeper and so on..
+- Lines of SQL should be no longer than 100 characters. TODO : How do we make sure of that? 
+- No trailing white space
+- If queries/models are run by DBT then do not mention schema or database name, just use table name
+- Use `!=` instead of `<>` since it is more popular in other programming languages
+- Ordering and grouping by a number (eg. GROUP BY 1, 2) is preferred
```
Comment (rr-yun-liaw) [inline README.md]: Actually I don't like ordering or grouping with numbers... 

I can only accept `group by 1`. Probably `group by 1, 2`. 

But often time I see people write `group by 1,2,3,4` or even `group by 1,3,5,7`. and it is super error prone and hard to read.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,462 @@
...
+- Commas seperating queries/CTE's should be at the end of each query/CTE instead of begining of next CTE, For example
+
+```
+-- Good
+with top10_interests as (
+
+    select distinct interest_area, interest_score, member_id 
+    from ff_last_year_member_interest_area
+
+    ),
```
Comment (rr-yun-liaw) [inline README.md]: Do you think these blanks lines and the 4 spaces before `)` are needed?

I would prefer

```sql
with top10_interests as (
    select distinct interest_area, interest_score, member_id
    from ff_last_year_member_interest_area
),
```

as the good example.

Comment (rr-yun-liaw) [reply README.md]: The blank lines can be added between CTEs, say,


```sql
with table_a as (
    ...
),
-- blank line here
with table_b as (
    ...
),
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,462 @@
...
+    from ff_last_year_member_interest_area
+    )
+
+, last_year_store_member_interest_area as (
+    select interest_area, rank as top10_rank 
+    ...
+    ...
+```
+
+- If you have to do nesting more than once or use multiple sub queries within a query, use CTEs instead. They are easy to read. Give meaningful names to them
```
Comment (rr-yun-liaw) [inline README.md:230]: Might add one thing that we suggest nested queries should not greater than 2 levels deep.

Comment (rr-manoj-dobbali) [reply README.md:230]: I meant the same by _If you have to do nesting more than once_

**Comment by rr-wei-lee** [comment]
Most of the existing rules do not conflict with https://docs.sqlfluff.com/en/stable/rules.html. I've made some changes through sqlfluff on  https://github.com/rr-manoj-dobbali/sql-style-guide/pull/2/commits/91f89d6b1b0f238a184e928972cfde498b829ed7

File: README.md (hunk 2 above)
Comment (rr-manoj-dobbali) [reply README.md]: I made this change

**Comment by rr-manoj-dobbali** [comment]
All the changes discussed and corrections are made

**Comment by rr-manoj-dobbali** [comment]
@rr-yun-liaw I think it is ready to be merged to master

**Review by rr-yun-liaw — APPROVED** [review]
LGTM. 

Once it is merged please post this in data science channel and asking for data scientists for comments. Also please mention that we are experiencing the tool SqlFluff to lint the SQLs.

**Pull request merged**